File: 019/019.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head, n):

        ## Input:
        # head: Optional[ListNode], n: int
        ## Output:
        # Optional[ListNode]
        
        p1, p2 = head, head
        for i in range(n):
            p1 = p1.next
        
        if (p1 == None):
            return head.next
        
        while (p1.next != None):
            p1 = p1.next
            p2 = p2.next

        p2.next = p2.next.next
        return head               

File: 019/019.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
    ListNode *removeNthFromEnd(ListNode *head, int n)
    {
        ListNode *p1 = head;
        ListNode *p2 = head;

        for (int i = 0; i < n; i++)
        {
            p1 = p1->next;
        }

        if (p1 == nullptr)
        {
            return head->next;
        }

        while (p1->next != nullptr)
        {
            p1 = p1->next;
            p2 = p2->next;
        }

        p2->next = p2->next->next;
        return head;
    }
};